clippy_lints/src/lifetimes_bound_nested_ref.rs

```diff
@@ -0,0 +1,222 @@
+        let mut declared_bounds = Vec::<BoundLifetimePair<'_>>::new();
+        for where_predicate in generics.predicates {
+            declared_bounds.extend(get_declared_bounds(*where_predicate));
+        }
+        // collect bounds implied by nested references with lifetimes in arguments
+        let mut implied_bounds = Vec::<BoundLifetimePair<'_>>::new();
+        for input_ty in fn_decl.inputs {
+            implied_bounds.extend(get_nested_ref_implied_bounds(input_ty));
+        }
+        // and for function return type
+        if let rustc_hir::FnRetTy::Return(ret_ty) = fn_decl.output {
+            implied_bounds.extend(get_nested_ref_implied_bounds(ret_ty));
+        }
+
+        let implied_bounds = implied_bounds;
+
+        for implied_bound in &implied_bounds {
+            if !declared_bounds.contains(implied_bound) {
+                span_lint(
+                    ctx,
+                    ADD_REDUNDANT_LIFETIMES_BOUND_NESTED_REF_FN,
+                    generics.span,
+                    &format!(
+                        "missing lifetime bound declation: {}",
+                        implied_bound.as_bound_declaration()
+                    ),
+                );
+            }
+        }
+
+        for declared_bound in &declared_bounds {
+            if implied_bounds.contains(declared_bound) {
+                span_lint(
+                    ctx,
+                    REMOVE_REDUNDANT_LIFETIMES_BOUND_NESTED_REF_FN,
+                    generics.span,
+                    &format!(
+                        "declared lifetime bound is implied: {}",
+                        declared_bound.as_bound_declaration()
+                    ),
+                );
+            }
+        }
+    }
+}
+
+fn get_declared_bounds<'a>(where_predicate: WherePredicate<'a>) -> Vec<BoundLifetimePair<'a>> {
+    let mut declared_bounds = Vec::new();
+    match where_predicate {
+        WherePredicate::BoundPredicate(_) | WherePredicate::EqPredicate(_) => {},
+        WherePredicate::RegionPredicate(where_region_predicate) => {
+            for generic_bound in where_region_predicate.bounds {
+                let GenericBound::Outlives(outlived_lifetime) = generic_bound else {
+                    continue;
+                };
+                let declared_bound_lifetime_pair = BoundLifetimePair {
+                    long_lifetime: where_region_predicate.lifetime,
+                    outlived_lifetime: outlived_lifetime,
+                };
+                declared_bounds.push(declared_bound_lifetime_pair);
+            }
+        },
+    }
+    declared_bounds
+}
+
+fn get_nested_ref_implied_bounds<'a>(ty: &Ty<'a>) -> Vec<BoundLifetimePair<'a>> {
+    let mut implied_bounds = Vec::new();
+    // collect only from top level reference
+    let TyKind::Ref(mut lifetime, mut mut_ty) = ty.kind else {
+        return implied_bounds;
+    };
+    while let TyKind::Ref(nested_lifetime, nested_mut_ty) = mut_ty.ty.kind {
```

Comment:
> also this doesn't handle type aliases. It may be easier to do this analysis on the `ty` datastructures instead of on the `hir` datastructures. You can get the `ty::FnSig` via `cx.tcx.fn_sig(local_def_id)`.

I'll try and use the ```ty::FnSig``` instead of the things from ```hir``` to collect the implied bounds. In the example ``` &'a (&'b i32, u32)``` I see that this will probably need recursion, and implementing that might take a while.